localstack/aws/connect.py

```diff
@@ -0,0 +1,473 @@
+        Return AWS region as set in the Boto session.
+        """
+        return self._session.region_name
+
+    def _get_region(self) -> str:
+        """
+        Return the AWS region name from following sources, in order of availability.
+        - LocalStack request context
+        - LocalStack default region
+        - Boto session
+        """
+        return (
+            get_region_from_request_context()
+            or config.DEFAULT_REGION
+            or self._get_session_region()  # TODO this will never be called, as DEFAULT_REGION defaults to 'us-east-1'
+        )
+
+
+class InternalClientFactory(ClientFactory):
+    def _get_client_post_hook(self, client: BaseClient) -> BaseClient:
+        """
+        Register handlers that enable internal data object transfer mechanism
+        for internal clients.
+        """
+        client.meta.events.register(
+            "provide-client-params.*.*", handler=_handler_create_request_parameters
+        )
+        client.meta.events.register("before-call.*.*", handler=_handler_inject_dto_header)
+
+        return client
+
+    def get_client(
+        self,
+        service_name: str,
+        region_name: Optional[str],
+        aws_access_key_id: Optional[str] = None,
+        aws_secret_access_key: Optional[str] = None,
+        aws_session_token: Optional[str] = None,
+        endpoint_url: str = None,
+        config: Config = None,
+    ) -> BaseClient:
+        """
+        Build and return client for connections originating within LocalStack.
+
+        All API operation methods (such as `.list_buckets()` or `.run_instances()`
+        take additional args that start with `_` prefix. These are used to pass
+        additional information to LocalStack server during internal calls.
+
+        :param service_name: Service to build the client for, eg. `s3`
+        :param region_name: Region name. See note above.
+            If set to None, loads from botocore session.
+        :param aws_access_key_id: Access key to use for the client.
+            Defaults to LocalStack internal credentials.
+        :param aws_secret_access_key: Secret key to use for the client.
+            Defaults to LocalStack internal credentials.
+        :param aws_session_token: Session token to use for the client.
+            Not being used if not set.
+        :param endpoint_url: Full endpoint URL to be used by the client.
+            Defaults to appropriate LocalStack endpoint.
+        :param config: Boto config for advanced use.
+        """
+
+        return self._get_client(
+            service_name=service_name,
+            region_name=region_name or self._get_region(),
+            use_ssl=self._use_ssl,
+            verify=self._verify,
+            endpoint_url=endpoint_url or get_local_service_url(service_name),
+            aws_access_key_id=aws_access_key_id or INTERNAL_AWS_ACCESS_KEY_ID,
+            aws_secret_access_key=aws_secret_access_key or INTERNAL_AWS_SECRET_ACCESS_KEY,
```

Comment:
The only difference in this function between the two factories is the fallback to the internal credentials here, right?

Reply:
Yes!